### DISCUSSION TITLE
Why does `jax.vmap` not support broadcasting along the mapped axes?

### DISCUSSION THREAD

**Discussion by adonath** [jax-ml/jax#29944] [Ideas] [answered]
I presume this has been discussed before but I could not find any reference, so I thought I bring it up as a question. I think in principle `vmap` could support broadcasting of the input arrays along the mapped axes. Consider the following minimal example:

```python
import jax
from jax import numpy as jnp

a = jax.random.normal(random.key(9876), shape=(5, 4, 10))
b = jax.random.normal(random.key(976), shape=(1, 10, 4))

# batched matmul
print(jnp.matmul(a, b).shape)

# batched via vmap and repeating b
print(jax.vmap(jnp.dot)(a, jnp.repeat(b, 5, axis=0)).shape)

# why not directly support broadcasting along the mapped axes?
jax.vmap(jnp.dot)(a, b)
```
I can see that `vmap` and other functions transforms are JAX specific and do not necessarily need to support Numpy concepts such as broadcasting, but in practice it would still be useful I think and conceptional still be well defined. But maybe I have overlooked something?

**Comment by jakevdp** [comment, marked as answer]
It's an interesting idea – fundamentally though, broadcasting is a NumPy concept. It's supported in JAX within the `jax.numpy` layer, but not in general supported in `jax.lax` or other JAX APIs. Implicit broadcasting is sometimes convenient, but for tools like `vmap` we think it's better and less error-prone to be explicit.

If you want something similar to `vmap` that is broadcast-aware, you can use [`jnp.vectorize`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.vectorize.html), which is implemented via `vmap` and does support broadcasting. In your example it would look something like this:
```python
result = jnp.vectorize(jnp.dot, signature="(a,b),(b,c)->(a,c)")(a, b)
print(result.shape)  # (5, 4, 4)
```

**Reply by jakevdp** [reply]
I should add: if you want a broadcasting `vmap` functionality it's easy enough to implement yourself. Here's an internal utility where we do this: https://github.com/jax-ml/jax/blob/ae7924995a3311299469bdebd3b2fb8461810bb0/jax/experimental/sparse/util.py#L60-L77

I wouldn't recommend importing this, as it's not part of the public API, and may be removed without warning. But it's about 20 lines and you can certainly define it locally if you'd like to use it.

**Reply by adonath** [reply]
Thanks @jakevdp! I was not aware of `jnp.vectorize` as a convenience / Numpy compatibility layer on top of `vmap`. Given that both exists I fully agree that the more general explicit behavior of `vmap` is very reasonable and does not need further convenience. I guess it makes sense to crosslink or have a short note in the `vmap` docstring to mention `jnp.vectorize` as well.

**Comment by Uripeer3** [comment]
@jakevdp Would you happen to know the performance takeaways? Speed is one thing, but I think I'm more concerned about VRAM usage and management, especially upon gradient calculation.
My use case is calculating the gradient of a complex multilayer wave propagation simulation, and I can implement some part of it broadcasted within the function or broadcast with vmap, trying to understand the consequences of this in the long run.

**Reply by jakevdp** [reply]
I would not anticipate any runtime difference between "broadcasted with vmap" and "broadcasted in the function". Both should lower to more-or-less the same HLO, and you should use whichever approach is more convenient to you.

Here's a quick example:
```python
import jax
import jax.numpy as jnp
import numpy as np

def f(M, x):
  """Function to operate on a single batch"""
  assert M.ndim == 2
  assert x.ndim == 1
  return jax.nn.silu(M @ x)

def f_batched(M, x):
  """Manually vectorized function"""
  assert M.ndim >= 2
  assert x.ndim >= 1
  return jax.nn.silu(jax.numpy.einsum("...ij,...j->...i", M, x))

# Automatically vectorized version using jnp.vectorize (jax.vmap under the hood)
f_vectorized = jnp.vectorize(f, signature="(a,b),(b)->(a)")

# Check that the two approaches produce equivalent results
M = np.random.randn(4, 5, 10, 10)
x = np.random.randn(4, 1, 10)
np.testing.assert_allclose(f_batched(M, x), f_vectorized(M, x))

# Check the jaxpr emitted by the two approaches
print(jax.make_jaxpr(f_batched)(M, x))
# { lambda ; a:f32[4,5,10,10] b:f32[4,1,10]. let
#     c:f32[4,10] = squeeze[dimensions=(1,)] b
#     d:f32[4,5,10] = dot_general[
#       dimension_numbers=(([3], [1]), ([0], [0]))
#       preferred_element_type=float32
#     ] a c
#     e:f32[4,5,10] = jit[
#       name=silu
#       jaxpr={ lambda ; d:f32[4,5,10]. let
#           f:f32[4,5,10] = logistic d
#           e:f32[4,5,10] = mul d f
#         in (e,) }
#     ] d
#   in (e,) }

print(jax.make_jaxpr(f_vectorized)(M, x))
# { lambda ; a:f32[4,5,10,10] b:f32[4,1,10]. let
#     c:f32[4,10] = squeeze[dimensions=(1,)] b
#     d:f32[4,5,10] = dot_general[
#       dimension_numbers=(([3], [1]), ([0], [0]))
#       preferred_element_type=float32
#     ] a c
#     e:f32[4,5,10] = jit[
#       name=silu
#       jaxpr={ lambda ; d:f32[4,5,10]. let
#           f:f32[4,5,10] = logistic d
#           e:f32[4,5,10] = mul d f
#         in (e,) }
#     ] d
#   in (e,) }
```
You can see that in this case the manually-batched and automatically vectorized versions lead to the exact same jaxpr, which implies they would have the exact same runtime characteristics. In more complicated situations there might be a difference between the two, but in general the design goal of `vmap` (which underlies `vectorize`) is to automatically generate batching code similar to what you might write by hand.
